File: backend/app/services/mock_timeseries.py
from __future__ import annotations

from datetime import date, timedelta

import numpy as np
import polars as pl


def _well_seed(well_id: str) -> int:
    return sum(ord(char) for char in well_id)


def _smooth_noise(
    rng: np.random.Generator,
    size: int,
    scale: float,
    window: int,
) -> np.ndarray:
    raw = rng.normal(0.0, scale, size + window - 1)
    kernel = np.ones(window, dtype=float) / window
    return np.convolve(raw, kernel, mode="valid")


def _smooth_step(timeline: np.ndarray, start: float, duration: float) -> np.ndarray:
    scaled = np.clip((timeline - start) / max(duration, 1e-6), 0.0, 1.0)
    return scaled * scaled * (3.0 - 2.0 * scaled)


def _gaussian_pulse(timeline: np.ndarray, center: float, sigma: float) -> np.ndarray:
    return np.exp(-0.5 * ((timeline - center) / max(sigma, 1e-6)) ** 2)


def _get_scenario_params(seed: int) -> dict[str, float]:
    variant = seed % 3

    scenarios: tuple[dict[str, float], ...] = (
        {
            "qliq_start": 124.0,
            "qliq_decline": 0.09,
            "degradation_start": 34.0,
            "degradation_duration": 38.0,
            "degradation_amp": 7.5,
            "downtime_center": 82.0,
            "downtime_sigma": 2.2,
            "downtime_amp": 20.0,
            "replacement_start": 86.0,
            "replacement_duration": 6.0,
            "replacement_amp": 8.0,
            "opz_start": 96.0,
            "opz_fade_start": 112.0,
            "opz_amp": 6.0,
            "water_start": 126.0,
            "water_duration": 34.0,
            "water_amp": 14.0,
            "regime_start": 148.0,
            "regime_duration": 12.0,
            "regime_amp": 2.2,
        },
        {
            "qliq_start": 118.0,
            "qliq_decline": 0.07,
            "degradation_start": 58.0,
            "degradation_duration": 22.0,
            "degradation_amp": 4.8,
            "downtime_center": 54.0,
            "downtime_sigma": 3.2,
            "downtime_amp": 10.5,
            "replacement_start": 103.0,
            "replacement_duration": 8.0,
            "replacement_amp": 5.5,
            "opz_start": 118.0,
            "opz_fade_start": 132.0,
            "opz_amp": 3.8,
            "water_start": 138.0,
            "water_duration": 24.0,
            "water_amp": 7.5,
            "regime_start": 74.0,
            "regime_duration": 10.0,
            "regime_amp": 3.0,
        },
        {
            "qliq_start": 121.0,
            "qliq_decline": 0.08,
            "degradation_start": 24.0,
            "degradation_duration": 30.0,
            "degradation_amp": 5.2,
            "downtime_center": 118.0,
            "downtime_sigma": 2.0,
            "downtime_amp": 8.0,
            "replacement_start": 122.0,
            "replacement_duration": 5.0,
            "replacement_amp": 4.0,
            "opz_start": 70.0,
            "opz_fade_start": 88.0,
            "opz_amp": 4.8,
            "water_start": 92.0,
            "water_duration": 42.0,
            "water_amp": 16.0,
            "regime_start": 146.0,
            "regime_duration": 14.0,
            "regime_amp": 3.6,
        },
    )

    return scenarios[variant]


def generate_well_timeseries(
    well_id: str,
    date_from: date | None = None,
    date_to: date | None = None,
) -> list[dict[str, object]]:
    total_days = 180
    end_date = date.today()
    start_date = end_date - timedelta(days=total_days - 1)

    dates = [start_date + timedelta(days=index) for index in range(total_days)]
    timeline = np.arange(total_days, dtype=float)
    seed = _well_seed(well_id)
    rng = np.random.default_rng(seed)
    params = _get_scenario_params(seed)

    degradation = _smooth_step(
        timeline,
        start=params["degradation_start"],
        duration=params["degradation_duration"],
    )
    downtime = _gaussian_pulse(
        timeline,
        center=params["downtime_center"],
        sigma=params["downtime_sigma"],
    )
    replacement_reset = _smooth_step(
        timeline,
        start=params["replacement_start"],
        duration=params["replacement_duration"],
    )
    opz_effect = np.where(
        timeline >= params["opz_start"],
        1.0 - np.exp(-0.18 * np.clip(timeline - params["opz_start"], 0.0, None)),
        0.0,
    ) - np.where(
        timeline >= params["opz_fade_start"],
        1.0 - np.exp(-0.08 * np.clip(timeline - params["opz_fade_start"], 0.0, None)),
        0.0,
    )
    opz_effect = np.clip(opz_effect, 0.0, None)
    water_breakthrough = _smooth_step(
        timeline,
        start=params["water_start"],
        duration=params["water_duration"],
    )
    regime_shift = _smooth_step(
        timeline,
        start=params["regime_start"],
        duration=params["regime_duration"],
    )

    qliq_trend = params["qliq_start"] - params["qliq_decline"] * timeline
    qliq = (
        qliq_trend
        - params["degradation_amp"] * degradation
        - params["downtime_amp"] * downtime
        + params["replacement_amp"] * replacement_reset
        + params["opz_amp"] * opz_effect
        - (0.32 * params["water_amp"]) * water_breakthrough
        - (0.9 * params["regime_amp"]) * regime_shift
        + 0.9 * np.sin(timeline / 17.0)
        + _smooth_noise(rng, total_days, scale=0.85, window=7)
    )
    qliq = np.clip(qliq, 62.0, None)

    water_cut = (
        23.0
        + 0.035 * timeline
        + 0.9 * degradation
        - 1.1 * opz_effect
        + params["water_amp"] * water_breakthrough
        + params["regime_amp"] * regime_shift
        + _smooth_noise(rng, total_days, scale=0.55, window=11)
    )
    water_cut = np.clip(water_cut, 8.0, 92.0)

    qoil = (
        qliq * (1.0 - water_cut / 100.0)
        - 1.8 * degradation
        + 2.2 * opz_effect
        - (2.2 + 0.08 * params["water_amp"]) * water_breakthrough
        + _smooth_noise(rng, total_days, scale=0.65, window=5)
    )
    qoil = np.minimum(qoil, qliq * 0.95)
    qoil = np.clip(qoil, 12.0, None)

    gas_factor = (
        198.0
        + 0.03 * timeline
        - 6.0 * degradation
        - 10.0 * downtime
        + 8.5 * replacement_reset
        + 14.0 * opz_effect
        + 12.0 * water_breakthrough
        + 5.0 * regime_shift
        + 4.0 * np.sin(timeline / 29.0)
        + _smooth_noise(rng, total_days, scale=3.2, window=9)
    )
    gas_factor = np.clip(gas_factor, 145.0, 285.0)

    qgas = qoil * gas_factor
    qgas = np.clip(qgas, 1800.0, None)

    gas_liquid_factor = qgas / np.maximum(qliq, 1.0)

    qliq_vfm = (
        qliq * (1.006 + 0.004 * np.sin(timeline / 23.0))
        + _smooth_noise(rng, total_days, scale=0.55, window=5)
    )
    qliq_vfm = np.clip(qliq_vfm, 60.0, None)

    intake_pressure = (
        116.5
        - 0.018 * timeline
        + (2.0 + 0.18 * params["degradation_amp"]) * degradation
        - (3.8 + 0.22 * params["downtime_amp"]) * downtime
        - (1.2 + 0.16 * params["replacement_amp"]) * replacement_reset
        + (1.8 + 0.14 * params["water_amp"]) * water_breakthrough
        + 0.7 * params["regime_amp"] * regime_shift
        + 1.1 * np.sin(timeline / 21.0)
        + _smooth_noise(rng, total_days, scale=0.75, window=7)
    )

    esp_frequency = (
        49.6
        - 0.18 * degradation
        - (1.0 + 0.12 * params["downtime_amp"]) * downtime
        + 0.14 * params["replacement_amp"] * replacement_reset
        - 0.025 * params["water_amp"] * water_breakthrough
        + 0.1 * params["regime_amp"] * regime_shift
        + _smooth_noise(rng, total_days, scale=0.08, window=9)
    )
    esp_frequency = np.clip(esp_frequency, 43.5, 52.5)

    load = (
        61.5
        - 0.025 * timeline
        - (0.9 + 0.12 * params["degradation_amp"]) * degradation
        - (2.8 + 0.18 * params["downtime_amp"]) * downtime
        + params["replacement_amp"] * 0.24 * replacement_reset
        + 0.2 * params["opz_amp"] * opz_effect
        + (0.4 + 0.08 * params["water_amp"]) * water_breakthrough
        + 0.9 * np.sin(timeline / 26.0)
        + _smooth_noise(rng, total_days, scale=0.45, window=7)
    )
    load = np.clip(load, 42.0, 74.0)

    frame = pl.DataFrame(
        {
            "date": dates,
            "qliq": np.round(qliq, 2),
            "qoil": np.round(qoil, 2),
            "qgas": np.round(qgas, 2),
            "gas_factor": np.round(gas_factor, 2),
            "gas_liquid_factor": np.round(gas_liquid_factor, 2),
            "qliq_vfm": np.round(qliq_vfm, 2),
            "water_cut": np.round(water_cut, 2),
            "intake_pressure": np.round(intake_pressure, 2),
            "esp_frequency": np.round(esp_frequency, 2),
            "load": np.round(load, 2),
        }
    )

    if date_from is not None:
        frame = frame.filter(pl.col("date") >= pl.lit(date_from))

    if date_to is not None:
        frame = frame.filter(pl.col("date") <= pl.lit(date_to))

    frame = frame.with_columns(pl.col("date").dt.strftime("%Y-%m-%d"))

    return frame.to_dicts()

File: frontend/src/services/mockTimeseries.ts
import type { TimeSeriesPoint } from '@/types/timeseries'

type MockTimeseriesParams = {
  date_from?: string
  date_to?: string
}

type ScenarioConfig = {
  qliqStart: number
  qliqDecline: number
  degradationStart: number
  degradationDuration: number
  degradationAmp: number
  downtimeCenter: number
  downtimeSigma: number
  downtimeAmp: number
  replacementStart: number
  replacementDuration: number
  replacementAmp: number
  opzStart: number
  opzFadeStart: number
  opzAmp: number
  waterStart: number
  waterDuration: number
  waterAmp: number
  regimeStart: number
  regimeDuration: number
  regimeAmp: number
}

const DEFAULT_TOTAL_DAYS = 180
const DEFAULT_END_DATE = '2025-06-30'
const DAY_MS = 86400000

function hashWellId(wellId: string): number {
  let hash = 2166136261

  for (let index = 0; index < wellId.length; index += 1) {
    hash ^= wellId.charCodeAt(index)
    hash = Math.imul(hash, 16777619)
  }

  return hash >>> 0
}

function createRng(seed: number): () => number {
  let state = seed || 1

  return () => {
    state += 0x6d2b79f5
    let next = state
    next = Math.imul(next ^ (next >>> 15), next | 1)
    next ^= next + Math.imul(next ^ (next >>> 7), next | 61)
    return ((next ^ (next >>> 14)) >>> 0) / 4294967296
  }
}

function round(value: number): number {
  return Number(value.toFixed(2))
}

function clamp(value: number, min: number, max?: number): number {
  if (typeof max === 'number') {
    return Math.min(Math.max(value, min), max)
  }

  return Math.max(value, min)
}

function isoToTimestamp(value: string): number {
  return new Date(`${value}T00:00:00Z`).getTime()
}

function timestampToIso(value: number): string {
  return new Date(value).toISOString().slice(0, 10)
}

function shiftIsoDate(value: string, deltaDays: number): string {
  return timestampToIso(isoToTimestamp(value) + deltaDays * DAY_MS)
}

function getDateBounds(params: MockTimeseriesParams): { dateFrom: string; dateTo: string } {
  const dateTo = params.date_to ?? DEFAULT_END_DATE
  const dateFrom =
    params.date_from ?? shiftIsoDate(dateTo, -(DEFAULT_TOTAL_DAYS - 1))

  if (isoToTimestamp(dateFrom) <= isoToTimestamp(dateTo)) {
    return { dateFrom, dateTo }
  }

  return { dateFrom: dateTo, dateTo: dateFrom }
}

function getScenario(seed: number): ScenarioConfig {
  const variant = seed % 3

  switch (variant) {
    case 0:
      return {
        qliqStart: 124,
        qliqDecline: 0.09,
        degradationStart: 0.2,
        degradationDuration: 0.22,
        degradationAmp: 7.5,
        downtimeCenter: 0.46,
        downtimeSigma: 0.02,
        downtimeAmp: 20,
        replacementStart: 0.48,
        replacementDuration: 0.05,
        replacementAmp: 8,
        opzStart: 0.53,
        opzFadeStart: 0.62,
        opzAmp: 6,
        waterStart: 0.7,
        waterDuration: 0.19,
        waterAmp: 14,
        regimeStart: 0.82,
        regimeDuration: 0.07,
        regimeAmp: 2.2
      }
    case 1:
      return {
        qliqStart: 118,
        qliqDecline: 0.07,
        degradationStart: 0.32,
        degradationDuration: 0.12,
        degradationAmp: 4.8,
        downtimeCenter: 0.3,
        downtimeSigma: 0.03,
        downtimeAmp: 10.5,
        replacementStart: 0.57,
        replacementDuration: 0.05,
        replacementAmp: 5.5,
        opzStart: 0.66,
        opzFadeStart: 0.74,
        opzAmp: 3.8,
        waterStart: 0.77,
        waterDuration: 0.14,
        waterAmp: 7.5,
        regimeStart: 0.4,
        regimeDuration: 0.06,
        regimeAmp: 3
      }
    default:
      return {
        qliqStart: 121,
        qliqDecline: 0.08,
        degradationStart: 0.12,
        degradationDuration: 0.17,
        degradationAmp: 5.2,
        downtimeCenter: 0.66,
        downtimeSigma: 0.02,
        downtimeAmp: 8,
        replacementStart: 0.68,
        replacementDuration: 0.04,
        replacementAmp: 4,
        opzStart: 0.39,
        opzFadeStart: 0.49,
        opzAmp: 4.8,
        waterStart: 0.51,
        waterDuration: 0.24,
        waterAmp: 16,
        regimeStart: 0.82,
        regimeDuration: 0.08,
        regimeAmp: 3.6
      }
  }
}

function smoothStep(position: number, start: number, duration: number): number {
  const scaled = clamp((position - start) / Math.max(duration, 1e-6), 0, 1)
  return scaled * scaled * (3 - 2 * scaled)
}

function gaussianPulse(position: number, center: number, sigma: number): number {
  const normalized = (position - center) / Math.max(sigma, 1e-6)
  return Math.exp(-0.5 * normalized * normalized)
}

function buildDates(dateFrom: string, dateTo: string): string[] {
  const dates: string[] = []
  let cursor = isoToTimestamp(dateFrom)
  const end = isoToTimestamp(dateTo)

  while (cursor <= end) {
    dates.push(timestampToIso(cursor))
    cursor += DAY_MS
  }

  return dates
}

export function generateMockTimeseries(
  wellId: string,
  params: MockTimeseriesParams = {}
): TimeSeriesPoint[] {
  const { dateFrom, dateTo } = getDateBounds(params)
  const dates = buildDates(dateFrom, dateTo)
  const seed = hashWellId(wellId)
  const scenario = getScenario(seed)
  const random = createRng(seed)

  return dates.map((date, index) => {
    const lastIndex = Math.max(dates.length - 1, 1)
    const progress = index / lastIndex

    const degradation = smoothStep(progress, scenario.degradationStart, scenario.degradationDuration)
    const downtime = gaussianPulse(progress, scenario.downtimeCenter, scenario.downtimeSigma)
    const replacementReset = smoothStep(progress, scenario.replacementStart, scenario.replacementDuration)
    const opzEffect = clamp(
      (progress >= scenario.opzStart
        ? 1 - Math.exp(-9 * (progress - scenario.opzStart))
        : 0) -
        (progress >= scenario.opzFadeStart
          ? 1 - Math.exp(-5 * (progress - scenario.opzFadeStart))
          : 0),
      0
    )
    const waterBreakthrough = smoothStep(progress, scenario.waterStart, scenario.waterDuration)
    const regimeShift = smoothStep(progress, scenario.regimeStart, scenario.regimeDuration)

    const qliqBase =
      scenario.qliqStart -
      scenario.qliqDecline * index -
      scenario.degradationAmp * degradation -
      scenario.downtimeAmp * downtime +
      scenario.replacementAmp * replacementReset +
      scenario.opzAmp * opzEffect -
      0.32 * scenario.waterAmp * waterBreakthrough -
      0.9 * scenario.regimeAmp * regimeShift +
      0.9 * Math.sin(index / 17) +
      (random() - 0.5) * 1.1
    const qliq = clamp(qliqBase, 62)

    const waterCut = clamp(
      23 +
        0.035 * index +
        0.9 * degradation -
        1.1 * opzEffect +
        scenario.waterAmp * waterBreakthrough +
        scenario.regimeAmp * regimeShift +
        (random() - 0.5) * 1.2,
      8,
      92
    )

    const qoil = clamp(
      Math.min(
        qliq * 0.95,
        qliq * (1 - waterCut / 100) -
          1.8 * degradation +
          2.2 * opzEffect -
          (2.2 + 0.08 * scenario.waterAmp) * waterBreakthrough +
          (random() - 0.5) * 1.3
      ),
      12
    )

    const gasFactor = clamp(
      198 +
        0.03 * index -
        6 * degradation -
        10 * downtime +
        8.5 * replacementReset +
        14 * opzEffect +
        12 * waterBreakthrough +
        5 * regimeShift +
        4 * Math.sin(index / 29) +
        (random() - 0.5) * 6.4,
      145,
      285
    )

    const qgas = clamp(qoil * gasFactor, 1800)
    const gasLiquidFactor = qgas / Math.max(qliq, 1)
    const qliqVfm = clamp(qliq * (1.006 + 0.004 * Math.sin(index / 23)) + (random() - 0.5) * 1.1, 60)
    const qliqWfm = clamp(qliq * (0.996 + 0.005 * Math.cos(index / 19)) + (random() - 0.5) * 0.9, 60)
    const intakePressure = round(
      116.5 -
        0.018 * index +
        (2 + 0.18 * scenario.degradationAmp) * degradation -
        (3.8 + 0.22 * scenario.downtimeAmp) * downtime -
        (1.2 + 0.16 * scenario.replacementAmp) * replacementReset +
        (1.8 + 0.14 * scenario.waterAmp) * waterBreakthrough +
        0.7 * scenario.regimeAmp * regimeShift +
        1.1 * Math.sin(index / 21) +
        (random() - 0.5) * 1.4
    )
    const espFrequency = clamp(
      49.6 -
        0.18 * degradation -
        (1 + 0.12 * scenario.downtimeAmp) * downtime +
        0.14 * scenario.replacementAmp * replacementReset -
        0.025 * scenario.waterAmp * waterBreakthrough +
        0.1 * scenario.regimeAmp * regimeShift +
        (random() - 0.5) * 0.16,
      43.5,
      52.5
    )
    const load = clamp(
      61.5 -
        0.025 * index -
        (0.9 + 0.12 * scenario.degradationAmp) * degradation -
        (2.8 + 0.18 * scenario.downtimeAmp) * downtime +
        scenario.replacementAmp * 0.24 * replacementReset +
        0.2 * scenario.opzAmp * opzEffect +
        (0.4 + 0.08 * scenario.waterAmp) * waterBreakthrough +
        0.9 * Math.sin(index / 26) +
        (random() - 0.5) * 0.9,
      42,
      74
    )

    return {
      date,
      qliq: round(qliq),
      qoil: round(qoil),
      qgas: round(qgas),
      gas_factor: round(gasFactor),
      gas_liquid_factor: round(gasLiquidFactor),
      qliq_wfm: round(qliqWfm),
      qliq_vfm: round(qliqVfm),
      water_cut: round(waterCut),
      intake_pressure: intakePressure,
      esp_frequency: round(espFrequency),
      load: round(load)
    }
  })
}
